xds/src/main/java/io/grpc/xds/XdsLoadBalancer2.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ChannelLogger.ChannelLogLevel;
+import io.grpc.ConnectivityState;
+import io.grpc.ConnectivityStateInfo;
+import io.grpc.LoadBalancer;
+import io.grpc.Status;
+import io.grpc.SynchronizationContext.ScheduledHandle;
+import io.grpc.util.ForwardingLoadBalancerHelper;
+import io.grpc.xds.XdsComms.AdsStreamCallback;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link LoadBalancer} that uses the XDS protocol.
+ *
+ * <p>This class manages fallback handling. The logic for child policy handling and fallback policy
+ * handling is provided by LookasideLb and FallbackLb.
+ */
+// TODO(zdapeng): migrate name to XdsLoadBlancer
+final class XdsLoadBalancer2 extends LoadBalancer {
+
+  private static final long FALLBACK_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(10); // same as grpclb
+
+  private final Helper helper;
+  private final LoadBalancer lookasideLb;
+  private final LoadBalancer.Factory fallbackLbFactory;
+  private final AdsStreamCallback adsCallback = new AdsStreamCallback() {
+    @Override
+    public void onWorking() {
+      if (childPolicyHasBeenReady) {
+        // cancel Fallback-After-Startup timer if there's any
+        cancelFallbackTimer();
+      }
+
+      childBalancerWorked = true;
+    }
+
+    @Override
+    public void onError() {
+      if (!childBalancerWorked) {
+        // start Fallback-at-Startup immediately
+        useFallbackPolicy();
+      } else if (childPolicyHasBeenReady) {
+        // TODO: schedule a timer for Fallback-After-Startup
+      } // else: the Fallback-at-Startup timer is still pending, noop and wait
+    }
+
+    @Override
+    public void onAllDrop() {
+      cancelFallback();
+    }
+  };
+
+  @Nullable
+  private LoadBalancer fallbackLb;
+  @Nullable
+  private ResolvedAddresses resolvedAddresses;
+  // Scheduled only once.  Never reset.
```

Comment:
It will be reset in the "Fallback-After-Startup" case, right? Since you already have stubs for "Fallback-After-Startup" in the code, the documentation should be consistent.

Reply:
Changed to `Never reset to null.`